File: src/main/java/com/leet/medium/FlattenBinaryTreeToLinkedList.go
package main

type TreeNode struct {
   Val int
   Left *TreeNode
   Right *TreeNode
}

func flattenUtil(node *TreeNode, tail *TreeNode) *TreeNode {
  if node == nil {
    return tail
  }
  right := flatten_2(node.Right, tail)
  left := flatten_2(node.Left, right)
  node.Left = nil
  node.Right = left
  return node
}

func flatten(root *TreeNode)  {
  flattenUtil(root, nil)
}

File: src/main/java/com/leet/medium/FlattenBinaryTreeToLinkedList.ts
function flattenUtil(node: TreeNode | null, tail:  TreeNode | null = null): TreeNode {
  if (node == null) return tail;
  const right = flattenUtil(node.right, tail);
  node.right = flattenUtil(node.left, right);
  node.left = null;
  return node;
}

function flatten(root: TreeNode | null): void {
  flattenUtil(root, null);
};
